### PR TITLE
feat(package-operator): add validation with webhook

### LINKED ISSUE #116: Dependent packages should not be uninstallable
A package that is a dependency of another package should not be uninstallable. This should be implemented via webhook validation. 

Note that checking owner references only will not be sufficient, because a package could have existed before the dependent package has been installed.

### LINKED ISSUE #322: Validate Package Updates via Webhook
It is not allowed to set a version on a `Package`, that is not compatible with versions that dependent packages require. This should be implemented as webhook.

### PR DESCRIPTION
**Pull request by kosmoz**
Closes #322  
Closes #116

## 📑 Description

This PR adds a validating admission webhook for the operator.

This webhook uses the existing DependencyManager.Validate function for validation.

For package deletion, it verifies that the package either doesn't have any owning packages, or that all owning packages have a deletion timestamp (this is to prevent a deadlock).

## ✅ Checks
- [x] My code requires changes to the documentation
- [ ] I have updated the documentation as required

## ℹ Additional Information

It is possible to run the operator on the host machine and use it for the validation webhook using an ExternalName service (only works on minikube). To run the operator including the webhook configuration, run `make install webhook cert run`.
- `install` deploys the CRDs
- `webhook` deploys the webhook config with a patch to target the host machine
- `cert` creates a self signed TLS certificate on the host machine and patches the webhook config with the CA bundle
- `run` executes the operator

When deployed, the TLS certificate is created/updated by a Job and CronJob running once per month in the glasskube-system namespace.

This PR is the continuation of #342

### CODE REVIEW

File: internal/webhook/package_validator.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+// to at least one other package. However, we do have to allow packages to be deleted if all their dependant
+// packages are already being deleted, because the webhook is also called if a package is garbage-collected and
+// this would lead to a dead lock otherwise:
+// The dependant package is not deleted because of the foreground deletion finalizer and the dependency is not
+// deleted because the dependant is not deleted.
+func (p *PackageValidatingWebhook) ValidateDelete(ctx context.Context, obj runtime.Object) (warnings admission.Warnings, err error) {
+	log := ctrl.LoggerFrom(ctx)
+	if pkg, ok := obj.(*v1alpha1.Package); ok {
+		log.Info("validate delete", "name", pkg.Name)
+		if dependants, err := p.OwnersOfType(&v1alpha1.Package{}, pkg); err != nil {
```
Comment (christophenne) [inline internal/webhook/package_validator.go:81]: I think this is missing the case in which there is a package dependent on `pkg` without having an owner reference on it.

**Review by christophenne — CHANGES_REQUESTED** [review]
Looking good – I tested it and it works as expected in most cases. As discussed I created #364 and found another issue at deletion (see comment).

File: internal/repo/client/fake/fakeclient.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+		f.Packages[name] = map[string]v1alpha1.PackageManifest{}
+	}
+	f.Packages[name][version] = manifest
+}
+
+func (f *FakeClient) Clear() {
+	f.Packages = map[string]map[string]v1alpha1.PackageManifest{}
+}
+
+var _ client.RepoClient = &FakeClient{}
```
[the remark this answers is not in the document]
Comment (kosmoz) [reply internal/repo/client/fake/fakeclient.go:30]: This tells the compiler that this type implements the RepoClient interface (which is not possible any other way :upside_down_face:)

File: internal/webhook/package_validator.go (hunk 1 above)
Comment (kosmoz) [reply internal/webhook/package_validator.go:81]: This is a good point. As discussed, we will implement this in a separate PR, as a similar functionality is also needed in other places.

**Pull request merged**